arangod/Aql/KShortestPathsNode.cpp

```diff
@@ -328,16 +337,60 @@ std::unique_ptr<ExecutionBlock> KShortestPathsNode::createBlock(
 #endif
 
   if (shortestPathType() == arangodb::graph::ShortestPathType::Type::KPaths) {
-    auto finder = std::make_unique<graph::KPathFinder>(*opts);
-    auto executorInfos =
-        KShortestPathsExecutorInfos(outputRegister, std::move(finder),
-                                    std::move(sourceInput), std::move(targetInput));
-    return std::make_unique<ExecutionBlockImpl<KShortestPathsExecutor<graph::KPathFinder>>>(
-        &engine, this, std::move(registerInfos), std::move(executorInfos));
+    if (opts->refactor()) {
+      // Create IndexAccessor for BaseProviderOptions (TODO: Location need to be
+      // changed in the future) create BaseProviderOptions
+      BaseProviderOptions forwardProviderOptions(opts->tmpVar(), buildUsedIndexes());
+      BaseProviderOptions backwardProviderOptions(opts->tmpVar(), buildReverseUsedIndexes());
+
+      arangodb::graph::TwoSidedEnumeratorOptions enumeratorOptions{opts->minDepth,
+                                                                   opts->maxDepth};
+
+      if (opts->query().queryOptions().getTraversalProfileLevel() ==
+          TraversalProfileLevel::None) {
+        using KPathRefactored =
+            TwoSidedEnumerator<FifoQueue<SingleServerProvider::Step>, PathStore<SingleServerProvider::Step>, SingleServerProvider>;
+        auto kPathUnique = std::make_unique<KPathRefactored>(
+            SingleServerProvider{opts->query(), forwardProviderOptions},
+            SingleServerProvider{opts->query(), backwardProviderOptions},
+            std::move(enumeratorOptions));
+
+        auto executorInfos =
+            KShortestPathsExecutorInfos(outputRegister, engine.getQuery(),
+                                        std::move(kPathUnique), std::move(sourceInput),
+                                        std::move(targetInput));
+        return std::make_unique<ExecutionBlockImpl<KShortestPathsExecutor<KPathRefactored>>>(
+            &engine, this, std::move(registerInfos), std::move(executorInfos));
+      } else {
+        using KPathRefactored =
+            TwoSidedEnumerator<QueueTracer<FifoQueue<SingleServerProvider::Step>>,
+                               PathStoreTracer<PathStore<SingleServerProvider::Step>>, ProviderTracer<SingleServerProvider>>;
+        // TODO: below copy paste from above. clean this up later.
```

Comment:
Any idea to do this properly? Using a local method? Or a better way available?

Reply:
I was thinking of a Factory, but we can add this later if we have more algorithms